File: Pacman-Game-OOP-master 2/EASendMail/Samples_ASPNetMvc/C#/WebProject1/WebProject1/Services/MassSendMailService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Threading;
using System.Threading.Tasks;
using WebProject1.Models;
using EASendMail;

namespace WebProject1.Services
{
 
    public class MassSendMailService
    {
        private static Dictionary<string, MassSendThreadPool> _threadPools = null;
        private static object _lock = null;

        public static void Init()
        {
            _lock = new object();
            _threadPools = new Dictionary<string, MassSendThreadPool>();
        }

        public static void CreateAsyncTask(MailTask mailTask)
        {
            lock (_lock)
            {
                var status = new MassSendMailStatus();
                status.TaskId = mailTask.TaskId;
                if (!_threadPools.ContainsKey(mailTask.TaskId))
                {
                    MassSendThreadPool threadPool = new MassSendThreadPool();
                    _threadPools.Add(mailTask.TaskId, threadPool);

                    Task.Run(() =>
                    {
                        threadPool.Start(mailTask);
                    });
                }
            }
        }

        public static MassSendMailStatus QueryStatus(string taskId)
        {
            MassSendMailStatus status = null;
            lock (_lock)
            {
                if (_threadPools.ContainsKey(taskId))
                {
                    var threadPool = _threadPools[taskId];
                    status = threadPool.Status();
                    if (status.Completed)
                    {
                        _threadPools.Remove(taskId);
                    }
                }
            }

            return status;
        }
    }
}

File: Pacman-Game-OOP-master 2/EASendMail/Samples_ASPNetMvc/VB/WebProject1/WebProject1/Services/MassSendMailService.vb
Imports System.Threading.Tasks

Public Class MassSendMailService
    Private Shared _threadPools As Dictionary(Of String, MassSendThreadPool) = Nothing
    Private Shared _lock As Object = Nothing

    Public Shared Sub Init()
        _lock = New Object()
        _threadPools = New Dictionary(Of String, MassSendThreadPool)()
    End Sub

    Public Shared Sub CreateAsyncTask(ByVal mailTask As MailTask)
        SyncLock _lock
            Dim status = New MassSendMailStatus()
            status.TaskId = mailTask.TaskId

            If Not _threadPools.ContainsKey(mailTask.TaskId) Then
                Dim threadPool As MassSendThreadPool = New MassSendThreadPool()
                _threadPools.Add(mailTask.TaskId, threadPool)
                Task.Run(Sub()
                             threadPool.Start(mailTask)
                         End Sub)
            End If
        End SyncLock
    End Sub

    Public Shared Function QueryStatus(ByVal taskId As String) As MassSendMailStatus
        Dim status As MassSendMailStatus = Nothing

        SyncLock _lock

            If _threadPools.ContainsKey(taskId) Then
                Dim threadPool = _threadPools(taskId)
                status = threadPool.Status()

                If status.Completed Then
                    _threadPools.Remove(taskId)
                End If
            End If
        End SyncLock

        Return status
    End Function
End Class